### ISSUE TITLE
Is there a documentation for how to configure restart chunkserver when network connection is lost?

### ISSUE THREAD

**Issue by scotthuang1989** [moosefs/moosefs#135]
I know this is a question for stackoverfow, but it seems stackoverflow is very inactive on this topic, so I figure this is the most promising place for question of moosefs.

except for this configuration,  a more specific documentation should be very helpful for new user to learn how to maintain this software.

**Comment by fchioralia** [comment]
In this you will find most of what you want ->
https://moosefs.com/Content/Downloads/moosefs-3-0-users-manual.pdf
Also in the cfg files, you can see what variables can be changed.
Regarding your problem: you don't need to do anything in case of a
connection problem, except fixing that problem. The master will see the
disconnected chunkserver and will put those chunks in endangered mode and
start replicating them to the assigned goal (I'm assuming that is at least
2, as best practices recommend it). When the network connection is up, the
master will see again the lost chunks and stops replication. Probably you
will end up with a few chunks over goal, depending of how long was the
connection problem and how you tweak master (for example in mfsmaster.cfg
you can change: MATOCS_TIMEOUT, REPLICATIONS_DELAY_INIT, etc.  ).
Hope it helps! Regards!

**Comment by xandrus** [comment]
Hi
I don't quite know what exactly you are asking for.
Bring the connection back and chunkserver will automatically start "talking" with MooseFS master.

If you want to prevent replicating chunks from disconnected chunkserver, set maintenance mode for such a chunkserver from CGI webpage or from the command line.

**Comment by scotthuang1989** [comment]
here is one of my issues:

network connection at our cluster is unstable last weekend, when I go to office at Monday, I found only 6 of 16 chunkserver is online,  I guess  moosefs automatically start replicate disconnected chunkserver, so my moosefs file system run out of disk space , all works depend on moosefs filesystem crashed.  

I manually restart chunkserver service, then all chunkserver back on line. 

**I want find a way of handling this situation autmatically: chunkserver can back on line when network is stable.**

@fchioralia  thanks for the advice, I will try to increase these configuration

MASTER_RECONNECTION_DELAY

but I don't find REPLICATIONS_DELAY_INIT in the user manual.

**Comment by xandrus** [comment]
Hi,
Thank you for this information.
You can find this parameter in mfsmaster.cfg file.

```
#initial delay in seconds before starting replications (default is 300)
REPLICATIONS_DELAY_INIT = 300
```

But this is an extreme case. I mean you lose over 50% of chunkservers.
How long this network outage takes?

If chunkservers weren't able to connect again to master server it means that mfschunkserver process was not running, or systemd stopped it after some numbers of retries.

Can you check this in your system logs?

journalctl -u moosefs-chunkserver

Or just system logs.
Debian OS family:
cat /var/log/syslog

RedHat/CentOS/FreeBSD OS family:
cat /var/log/messages

**Comment by scotthuang1989** [comment]
It seems the logs are lost. I use this command, but it still only give me today's log.

journalctl -u moosefs-chunkserver -S 2018-08-25

Is this the default setting? If not,  maybe it is because there is an electricity outage in the city  because of thunderstorm this morning. 

I will get the log when there are chunkservers went off line again. 

I upload the logs of today, it have some time-out event,  not sure if this is normal.

[chunk_log.log](https://github.com/moosefs/moosefs/files/2331507/chunk_log.log)

**Comment by borkd** [comment]
> I want find a way of handling this situation autmatically: chunkserver can back on line when  network is stable.

- Scenario 1: natural disaster wiped out a datacenter with a subset of your chunkservers, there is no chance they are ever coming back
- Scenario 2: a prolonged network outage happened anywhere upstream of said subset of your chunkservers

Both will look as a network problem, and there is **no** way moosefs, or any other **filesystem can distinguish between the two during the outage purely on their own**.

Let's make it easier and assume you are around to witness the network outage. Do you up the replication limits while checking your account balance prior to adding new capacity elsewhere, or do you just make a temporary flip into maintenance mode? 

If the data is important enough you will **use external sources of information** (from out-of-band monitoring, automation, news feeds, etc..) to help make that call. **In the meantime**, while you are working on stitching those to work while nobody is watching, make sure to **provision enough capacity** to handle large network partitions, and follow https://github.com/moosefs/moosefs/issues/133.

As you think "_Screw this. I will just deploy this in the cloud._", or just need a laugh, read https://aphyr.com/posts/288-the-network-is-reliable

**Comment by scotthuang1989** [comment]
@borkd , thanks for the information. But my problems now is the chunkserver con't back on line even after network is stable.  I am trying to monitor the chunkserver service and restart it if it crash, but I want know if moosefs have built-in functionality to deal it it.

**Comment by richarson** [comment]
@scotthuang1989 if you're using a systemd distro and the mfschunkserver process crashes, you can try the 'Restart=on-failure' option for systemd service units.

Though I see that, at least on CentOS 7, it already has 'Restart=on-abnormal' set, but you might want to change it to 'on-failure':

```
[root@chunkserver ~] # systemctl cat moosefs-chunkserver.service
# /usr/lib/systemd/system/moosefs-chunkserver.service
[Unit]
Description=MooseFS Chunkserver
Wants=network-online.target
After=network.target network-online.target

[Service]
Type=forking
ExecStart=/usr/sbin/mfschunkserver start
ExecStop=/usr/sbin/mfschunkserver stop
ExecReload=/usr/sbin/mfschunkserver reload
PIDFile=/var/lib/mfs/.mfschunkserver.lock
Restart=on-abnormal

[Install]
WantedBy=multi-user.target
```

**Comment by scotthuang1989** [comment]
@richarson  thanks. one of my server crash this morning, and log shows that the reason is "master can't reach", I apply your solution and see if it fix the problem.

**Comment by scotthuang1989** [comment]
@richarson change Restart to 'on-failure' didn't work. one of my chunkserver crash again. and I had to restart serverice manually to bring it back.

**Comment by xandrus** [comment]
@scotthuang1989 Are we talking about real software crush?
Would you be so kind end send us the system logs or journalctl -u moosefs-chunkserver info?

**Comment by scotthuang1989** [comment]
what do you mean "real software crush". our company setup a small cluster by our own. none of us is professional system admin. so some machine will crush with "no reason"(we use chunkserver for our training task, So maybe there are resource isssues),  I can't get a full log for this error, I will send out once I have it.

**Comment by scotthuang1989** [comment]
I think  I found cause of this issue. after tracking this issue for weeks, I found that this is caused by resource conflict, the server we setup chunk service on , we also run computation work on it. some time chunkserver can't get resource it need,  if I try to restart chunkserver when CPU is very busy, it will fail.

**Comment by borkd** [comment]
@scotthuang1989 - for the sake of documenting workarounds for issues people had with their deployments - how are you dealing with this problem since root cause was discovered?

**Comment by scotthuang1989** [comment]
@borkd , I found that on some of my servers. process **mfschunkserver** will take a lot CPU resources, I **still don't know why it act this way, because according to documentation, the requirement configuration for chunkserver  is pretty low**.   my cpu model is: Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz

My workaround is removing some of chunkserver, I lose some capacity, but the file system is much more stable now.

**Comment by borkd** [comment]
@scotthuang1989 - there is at least one scenario when spike in the load on the chunkserver is to be expected: when chunkserver starts scanning disks for chunks with cold caches, and the underlying filesystem uses compression. I've had close to 200 chunkservers and never had timeouts during cold startup. 

If there is a regression, it would nice to have this behavior replicated. When you have a chance, please paste the graphs with cpu and memory load from CGI both for master and the chunkserver in question.

**Comment by scotthuang1989** [comment]
@borkd, you mean take s screen shot of CGI (web browser page?) or  output of top command?

**Comment by borkd** [comment]
CGI should suffice, but to put things in perspective you could attach output from other monitoring tools you may have in place which would show the conditions around timeout/disconnect.

**Comment by xandrus** [comment]
Please check if your hard disks are working correctly and there are no errors on the filesystem.
By the way, chunkserver CPU is close to 2.2 points :)

Basically, your cluster performs a lot of deletions. You are deleting over 5 million files. This operation will affect cluster performance.
Please adjust these parameters in your /etc/mfs/mfsmaster.cfg file

These values are default so I suggest to decrease them to 5 and 12, reload master server and check situation again.
CHUNKS_SOFT_DEL_LIMIT = 10
CHUNKS_HARD_DEL_LIMIT = 25

**Comment by scotthuang1989** [comment]
thanks. I will try that and try to get a more stable cpu utilization of chunkserver . 
BTW, if I want to free up spaces and I am sure I don't want recover the deleted files, it there a dirty way of deleting deleted files.

**Comment by borkd** [comment]
@scotthuang1989 - as @xandrus pointed out, your bottleneck seems to be removal of substantial number of chunks. MooseFS works best as a scale-out filesystem - when operated as such this would be much harder to run into. With that said, one can tweak MooseFS to operate at much smaller scale, but should not expect wonders when only a handful of consumer grade SATA disks is backing lifecycle changes in millions of chunks.

Easiest: Assuming you have spare capacity, and your system is not in 24x7 use, in addition to reducing DEL limits consider extending trash time past the computation phase.

Then look into the chunkserver machines:
- number of devices and performance of underlying storage dedicated to moosefs
- per device IO scheduler and queue depth configuration
- underlying filesystem performance when dealing with large number of (small) file operations
- IO priority tweaks
- bottlenecks in network configuration which may appear as rebalancing and replication competes with client traffic and control traffic to and from master

**Comment by scotthuang1989** [comment]
@borkd . ok, got it. This moosefs is kind of my private storage, company have set up a dedicated hdfs server and I think it's not convenient for my use case, so for now, I can't put more resources on this.  I will try to tweak moosefs to make some trade-off between performance and reliability.
